### PR TITLE
feat: validation and defaut func

### PR DESCRIPTION
**Pull request by kindermoumoute**
* feat: validation func for UUID
* feat: validation func for zone
* feat: validation func for region
* feat: default func for random name

### CODE REVIEW

File: scaleway/validation.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func validationZone() func(interface{}, string) ([]string, []error) {
+	return func(v interface{}, key string) (warnings []string, errors []error) {
+		rawZone, isString := v.(string)
+		if !isString {
+			return nil, []error{fmt.Errorf("invalid zone: not a string")}
+		}
+
+		zone, _ := utils.ParseZone(rawZone)
```
Comment (QuentinBrosse) [inline scaleway/validation.go]: Add a `// TODO: Use utils.ParseZone when the format validation will be implemented.` ?

File: scaleway/validation.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func validationRegion() func(interface{}, string) ([]string, []error) {
+	return func(v interface{}, key string) (warnings []string, errors []error) {
+		rawRegion, isString := v.(string)
+		if !isString {
+			return nil, []error{fmt.Errorf("invalid region: not a string")}
+		}
+
+		region, _ := utils.ParseRegion(rawRegion)
```
Comment (QuentinBrosse) [inline scaleway/validation.go]: Add a `// TODO: Use utils.ParseRegion when the format validation will be implemented.`?

File: scaleway/validation.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+		if rawRegion == "par1" || rawRegion == "ams1" {
+			warnings = append(warnings, fmt.Sprintf("%s is a deprecated name for region, use %v instead", rawRegion, region))
+		} else if !region.Exists() {
+			warnings = append(warnings, "%s region does not exist", rawRegion)
```
Comment (QuentinBrosse) [inline scaleway/validation.go]: "%s region does not _recognized_"?

File: scaleway/helpers.go [hunk 4]
```diff
@@ -310,37 +310,46 @@ func is403Error(err error) bool {
...
 }
 
 // getRandomName returns a random name prefixed for terraform.
 func getRandomName(prefix string) string {
 	return namegenerator.GetRandomName("tf", prefix)
 }
+
+func defaultFuncRandomName(prefix string) func() (interface{}, error) {
+	return func() (interface{}, error) {
+		return getRandomName(prefix), nil
```
Comment (jerome-quere) [inline scaleway/helpers.go]: should it add `tf` has pre-prefix ?

File: scaleway/validation.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	"github.com/scaleway/scaleway-sdk-go/utils"
+)
+
+// validationUUID validate the schema follows the canonical UUID format
```
Comment (DenBeke) [inline scaleway/validation.go]: validationUUID validates the schema following the canonical UUID format

File: scaleway/validation.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+func validationZone() func(interface{}, string) ([]string, []error) {
+	return func(v interface{}, key string) (warnings []string, errors []error) {
+		rawZone, isString := v.(string)
+		if !isString {
+			return nil, []error{fmt.Errorf("invalid zone: not a string")}
+		}
+
+		zone, _ := utils.ParseZone(rawZone)
+		if rawZone == "par1" || rawZone == "ams1" {
+			warnings = append(warnings, fmt.Sprintf("%s is a deprecated name for zone, use %v instead", rawZone, zone))
```
Comment (jerome-quere) [inline scaleway/validation.go]: I hope this will not print 3000000000 warning if the function is call many times

File: scaleway/validation.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+// "6ba7b810-9dad-11d1-80b4-00c04fd430c8".
+func validationUUID() func(interface{}, string) ([]string, []error) {
+	return func(v interface{}, key string) (warnings []string, errors []error) {
+		uuid, isString := v.(string)
+		if !isString {
+			return nil, []error{fmt.Errorf("invalid UUID: not a string")}
+		}
+
+		t := []byte(uuid)
+		if len(t) != 36 || t[8] != '-' || t[13] != '-' || t[18] != '-' || t[23] != '-' {
```
Comment (DenBeke) [inline scaleway/validation.go]: Wouldn't it better to use a uuid parsing lib here?

File: scaleway/helpers.go (hunk 4 above)
Comment (kindermoumoute) [reply scaleway/helpers.go]: getRandomName already add the `tf` prefix

File: scaleway/validation.go (hunk 3 above)
Comment (DenBeke) [reply scaleway/validation.go]: *is* in stead of *does* in that case 😜

File: scaleway/validation.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+		rawZone, isString := v.(string)
+		if !isString {
+			return nil, []error{fmt.Errorf("invalid zone: not a string")}
+		}
+
+		zone, _ := utils.ParseZone(rawZone)
+		if rawZone == "par1" || rawZone == "ams1" {
+			warnings = append(warnings, fmt.Sprintf("%s is a deprecated name for zone, use %v instead", rawZone, zone))
+		} else if !zone.Exists() {
+			warnings = append(warnings, "%s zone does not exist", rawZone)
```
[the remark this answers is not in the document]
Comment (DenBeke) [reply scaleway/validation.go]: *is* instead of *does* in that case 😜

File: scaleway/validation.go (hunk 6 above)
Comment (kindermoumoute) [reply scaleway/validation.go]: One warning 
image

**Comment by kindermoumoute** [comment]
```terraform
provider "scaleway" {
  project_id = "<YOUR-ORGANIZATION-ID>"
  secret_key = "<YOUR-SECRET-TOKEN>"
  zone       = "ams1"
}

resource "scaleway_compute_instance_ip" "ip" {
  zone       = "par1"
  project_id = "<YOUR-ORGANIZATION-ID>"
}

```

The output I have with the above config:

image

File: scaleway/validation.go (hunk 7 above)
Comment (kindermoumoute) [reply scaleway/validation.go]: Didn't want to add a dep for one function.

File: scaleway/helpers.go [hunk 9]
```diff
@@ -310,37 +310,46 @@ func is403Error(err error) bool {
...
+		Computed:     true,
+		ValidateFunc: validationRegion(),
 	}
 }
 
 // getRandomName returns a random name prefixed for terraform.
 func getRandomName(prefix string) string {
 	return namegenerator.GetRandomName("tf", prefix)
 }
+
```
Comment (QuentinBrosse) [inline scaleway/helpers.go]: Missing godoc

**Pull request merged**